tests/test_auto_tag.py

```
@@ -73,6 +73,52 @@ def tagger(url):
         assert supabase.table.return_value.insert.call_count == 1
 
+    def test_hung_image_does_not_block_batch(self):
+        """A stalled Gemini call on one image must not wedge the worker.
+
+        Simulates a hung tag call (via asyncio.sleep longer than the timeout)
+        and asserts the batch still completes, the stuck image is recorded
+        as a failure, and the healthy images succeed.
+        """
+        supabase = _make_client_with_insert_id()
+
+        async def fake_tag_and_insert(client, user_id, image_url, semaphore):
```

Ruff (0.15.12)

[warning] Missing return type annotation for private function `fake_tag_and_insert`

(ANN202)